[ISSUE-815] Generate Random Numbers Asynchronously on the GPU

AndrewBMadison opened the pull request:
<!-- Link to the issue and use the appropriate closing keyword (e.g., Closes, Resolves) -->
Closes #815

<!-- Please provide a brief overview of the changes implemented -->
#### Description
Replaced the custom Mersenne Twister GPU kernel with an AsyncPhilox_d class that asynchronously fills GPU buffers with random noise using cuRAND's Philox generator. The class supports double-buffering and is designed for concurrent execution.

GPUModel initializes Philox states and fills two initial buffers via loadPhilox() on a member AsyncPhilox_d instance. During each advance() call, requestSegment() retrieves a float* slice from the currently active buffer, sized appropriately for each vertex and ready to be used in advanceVertices(). 

Once a buffer is consumed, fillBuffer() is triggered on the other buffer while the current one continues to serve slices. This ensures continuous data availability through double-buffering.

AsyncPhilox_d uses its own internal CUDA stream to launch fill kernels asynchronously. To enable true concurrency, all other compute kernels needed to also use non-default streams. This is necessary because stream 0 (the default stream) implicitly synchronizes with all other streams, preventing concurrent execution and causing the scheduler to serialize kernel launches even when they could run in parallel.

<!-- Please check the boxes as applicable -->
#### Checklist (Mandatory for new features)
- [ ] Added Documentation 
- [ ] Added Unit Tests 

<!-- Ensure all boxes are checked before submitting the PR -->
#### Testing (Mandatory for all changes)
- [x] GPU Test: `test-medium-connected.xml` Passed
- [x] GPU Test: `test-large-long.xml` Passed

<!-- 
PR Guidelines
1. Ensure that your changes are merged into the `development` branch. Only merge into the `master` branch if explicitly instructed.
     - On GitHub, at the top of the PR page, change the base branch from `master` to `development`.
2. Assign the PR to yourself and apply the appropriate labels.
3. Only add a reviewer after submitting the PR and confirming that all GitHub Actions have passed.

Thank you for your contribution!
-->

AndrewBMadison committed:
building a new class for gpu rng

AndrewBMadison committed:
refactor

AndrewBMadison committed:
curand rng class

AndrewBMadison committed:
Merge remote-tracking branch 'origin/SharedDevelopment' into issue-815-async-rng

AndrewBMadison committed:
revert change

AndrewBMadison committed:
fix performance metrics

AndrewBMadison committed:
Testing refactor

AndrewBMadison committed:
fix performance metrics bug

AndrewBMadison committed:
added logs and increased size of buffers

AndrewBMadison committed:
testing various gpu memory allocation sizes

AndrewBMadison committed:
added include paths for CUDA include in cmakelists to fix a bug trying to #include <cuda_runtime.h>

AndrewBMadison committed:
testing

AndrewBMadison committed:
testing full simulation and adding in error handling

AndrewBMadison committed:
changed rng to philox and performed an analysis on the distribution

AndrewBMadison committed:
turned off file logging and fixed double free bug

AndrewBMadison committed:
fixed non-blocking stream

AndrewBMadison committed:
improved nvtx support

AndrewBMadison committed:
clang-format

AndrewBMadison committed:
Merge branch 'SharedDevelopment' into issue-815-async-rng

AndrewBMadison committed:
comments and renaming

AndrewBMadison committed:
Merge branch 'issue-815-async-rng' into AndrewDevelopment

AndrewBMadison committed:
clang

AndrewBMadison committed:
cleaned up changes

AndrewBMadison committed:
clang

AndrewBMadison committed:
format

AndrewBMadison committed:
format

AndrewBMadison committed:
fixed cpp version of utils library and added appropriate include for Book.h

AndrewBMadison committed:
revert clang changes on MersenneTwister

AndrewBMadison committed:
fixing cpu errors with cuda includes

AndrewBMadison committed:
cleaned up linking and unnecessary includes

AndrewBMadison committed:
clang format issues

AndrewBMadison committed:
clang format issues

AndrewBMadison committed:
removed unnecessary file removal

stiber commented on Simulator/Connections/Connections.h:
```diff
@@ -87,7 +92,8 @@ class Connections {
    ///  @param  layout               Layout information of the graph network.
```
Please update the comment for this method to reflect current set of parameters.

stiber commented on Simulator/Connections/Neuro/ConnGrowth.h:
```diff
@@ -127,8 +127,9 @@ class ConnGrowth : public Connections {
    ///  @param  layout               The Layout object.
```
Please update the comment for this method to reflect current set of parameters.

stiber commented on Simulator/Edges/AllEdges.cpp:
```diff
@@ -217,9 +217,16 @@ void AllEdges::createEdgeIndexMap(EdgeIndexMap &edgeIndexMap)
 }
 
+#if defined(USE_GPU)
+// Set GPU stream for edge kernels
```
Document a bit better in comment.

stiber commented on Simulator/Utils/RNG/MersenneTwister_d.h:
Is this still used?

AndrewBMadison replied:
It is no longer used, but I had left is in in case someone wanted to uncomment the code that uses the old RNG and test it. If you would like I could remove these files.

stiber commented on Simulator/Vertices/AllVertices.cpp:
```diff
@@ -49,3 +49,12 @@ void AllVertices::loadEpochInputs(uint64_t currentStep, uint64_t endStep)
+// Set Cuda Stream for Vertices kernels
+void AllVertices::SetStream(cudaStream_t pStream)
+{
+   stream = pStream;
```
`stream` is a data member of the class; should be named `stream_`. Is this supposed to be distinct from the `AllEdges` stream?

AndrewBMadison replied:
The new name for the stream is simulationStream and it is the same stream used by all synchronous kernels.

stiber commented on Simulator/Vertices/AllVertices.h:
```diff
@@ -92,7 +95,13 @@ class AllVertices {
 
 #if defined(USE_GPU)
+   ///  Cuda Stream for Vertices Kernels
+   cudaStream_t stream;
```
`stream` is a data member of the class; should be named `stream_`. Is this supposed to be distinct from the `AllEdges` stream?

AndrewBMadison replied:
The new name for the stream is simulationStream and it is the same stream used by all synchronous kernels.

stiber commented on Simulator/Vertices/Neuro/AllIZHNeurons_d.cpp:
```diff
@@ -171,7 +171,7 @@ void AllIZHNeurons::advanceVertices(AllEdges &synapses, void *allVerticesDevice,
 
    // Advance neurons ------------->
-   advanceIZHNeuronsDevice<<<blocksPerGrid, threadsPerBlock>>>(
+   advanceIZHNeuronsDevice<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(
```
`stream` is a data member of the class; should be named `stream_`. Is this supposed to be distinct from the `AllEdges` stream?

stiber reviewed (changes requested):
Minor cleanup.

AndrewBMadison committed:
Wrote better comments with doxygen in mind and renamed variables according to the code style

AndrewBMadison committed:
clang

AndrewBMadison committed:
Added documentation

AndrewBMadison committed:
fix nvtx variable names

AndrewBMadison commented on Simulator/Edges/Neuro/AllSTDPSynapses_d.cpp:
```diff
@@ -268,7 +268,7 @@ void AllSTDPSynapses::advanceEdges(void *allEdgesDevice, void *allVerticesDevice
    int blocksPerGrid = (totalEdgeCount_ + threadsPerBlock - 1) / threadsPerBlock;
    // Advance synapses ------------->
-   advanceSTDPSynapsesDevice<<<blocksPerGrid, threadsPerBlock>>>(
+   advanceSTDPSynapsesDevice<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(
```
The new name for the stream is simulationStream and it is the same stream used by all synchronous kernels.

AndrewBMadison commented on Simulator/Vertices/Neuro/AllSpikingNeurons_d.cpp:
```diff
@@ -194,7 +194,7 @@ void AllSpikingNeurons::integrateVertexInputs(void *allVerticesDevice,
    int vertex_count = Simulator::getInstance().getTotalVertices();
 
-   calcSummationPointDevice<<<blocksPerGrid, threadsPerBlock>>>(
+   calcSummationPointDevice<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(
```
The new name for the stream is simulationStream and it is the same stream used by all synchronous kernels.

stiber reviewed (dismissed):
Looks great; I will merge this.

stiber committed:
Merge branch 'SharedDevelopment' into AndrewDevelopment

stiber commented:
I take it back; this needs to have SharedDevelopment merged into it. There may be a conflict with the changes to device memory allocation/deallocation being moved to `OperationManager`. We will need to let all the tests run, plus I will add a comment to one location in the code.

stiber commented on Simulator/Core/GPUModel.cpp:
```diff
@@ -53,28 +53,31 @@ void GPUModel::allocDeviceStruct()
-   HANDLE_ERROR(cudaFree(synapseIMapDevice.outgoingEdgeIndexMap_));
-   HANDLE_ERROR(cudaFree(synapseIMapDevice.incomingEdgeBegin_));
-   HANDLE_ERROR(cudaFree(synapseIMapDevice.incomingEdgeCount_));
-   HANDLE_ERROR(cudaFree(synapseIMapDevice.incomingEdgeIndexMap_));
-   HANDLE_ERROR(cudaFree(edgeIndexMapDevice_));
-   HANDLE_ERROR(cudaFree(randNoise_d));
+   vertices.deleteVerticesDeviceStruct(*allVerticesDevice);
+   // Copy device edge and vertex structs to host memory
+   edges.copyEdgeDeviceToHost(*allEdgesDevice);
+   // Deallocate device memory
+   edges.deleteEdgeDeviceStruct(*allEdgesDevice);
+   // HANDLE_ERROR(cudaFree(randNoise_d));
+   // closeFileMT();
+   HANDLE_ERROR(cudaStreamDestroy(simulationStream_));
```
This method needs review. Questions include:
1. Where are `vertices` and `edges` declared?
2. Given the new `OperationManager` control that has all classes managing their own device memory, are any of these "non-local" deletions needed?

stiber commented on Simulator/Core/GPUModel.cpp:
```diff
@@ -124,8 +133,11 @@ void GPUModel::finish()
    OperationManager::getInstance().executeOperation(Operations::copyFromGPU);
    // deallocates memories on CUDA device
+   // TODO: Resolve whether the deletion operations below that don't seem bundled
+   // with the OperationManager are necessary, should be moved, or if they can be removed.
+   AsyncGenerator_.deleteDeviceStruct();
    OperationManager::getInstance().executeOperation(Operations::deallocateGPUMemory);
-
+   deleteEdgeIndexMap();
```
As the TODO says, the two lines that delete `AsyncGenerator_` and the edge index map either need to be moved or aren't necessary.

AndrewBMadison replied:
I have moved the deletion of AsyncGenerator_ to deleteDeviceStruct(), and other line is now deleted from the remerge of SharedDevelopment.

stiber commented on Simulator/Core/GPUModel.cpp:
```diff
@@ -238,7 +251,8 @@ void GPUModel::updateConnections()
+                                       simulationStream_);
       // create edge index map
       connections_->createEdgeIndexMap();
       // copy index map to the device memory
```
Should the line below use `OperationManager`, or do we want finer granularity?

AndrewBMadison replied:
That is a good question and it looks like OperationManager could be used to execute this operation, but I don't know if I'm the right person to answer that question.

stiber reviewed (changes requested):
Besides the comments below, need to examine `GPUModel::allocEdgeIndexMap()` and `GPUModel::copyCPUtoGPU()` to see if they need rewrites because of `DeviceVector`

AndrewBMadison committed:
Revert "Merge branch 'SharedDevelopment' into AndrewDevelopment"

This reverts commit 5d7ee8c1ce9aaa32ca6c49f87969b6b4bb64ecda, reversing
changes made to 7aa77eba96202d9097d140758f4fda8ec2019c87.

AndrewBMadison committed:
Manually remerged SharedDevelopment into AndrewDevelopment by applying the diff and reviewing changes to resolve conflicts.

AndrewBMadison committed:
moved deallocation of AsyncGenerator to deleteDeviceStruct

AndrewBMadison commented:
I reverted and remerged SharedDevelopment into AndrewDevelopment, manually reviewing changes and deleting old unnecessary code from the SharedDevelopment commit before Ben merged his code in. This resolved a lot of the changes you requested, but I am unsure if the OperationManager should execute the copyCPUtoGPU so maybe you could ask Ben. I also moved the AsyncGenerator deletion as you requested.
